Source/Scene/ShadowMap.js

```diff
@@ -0,0 +1,397 @@
+/*global define*/
+define([
+        '../Core/BoundingRectangle',
+        '../Core/BoxOutlineGeometry',
+        '../Core/Cartesian3',
+        '../Core/Color',
+        '../Core/ColorGeometryInstanceAttribute',
+        '../Core/defined',
+        '../Core/destroyObject',
+        '../Core/GeometryInstance',
+        '../Core/Matrix4',
+        '../Core/PixelFormat',
+        '../Core/SphereOutlineGeometry',
+        '../Renderer/ClearCommand',
+        '../Renderer/Framebuffer',
+        '../Renderer/PassState',
+        '../Renderer/PixelDatatype',
+        '../Renderer/RenderState',
+        '../Renderer/ShaderProgram',
+        '../Renderer/ShaderSource',
+        '../Renderer/Texture',
+        './Camera',
+        './CullFace',
+        './OrthographicFrustum',
+        './Pass',
+        './PerInstanceColorAppearance',
+        './Primitive'
+    ], function(
+        BoundingRectangle,
+        BoxOutlineGeometry,
+        Cartesian3,
+        Color,
+        ColorGeometryInstanceAttribute,
+        defined,
+        destroyObject,
+        GeometryInstance,
+        Matrix4,
+        PixelFormat,
+        SphereOutlineGeometry,
+        ClearCommand,
+        Framebuffer,
+        PassState,
+        PixelDatatype,
+        RenderState,
+        ShaderProgram,
+        ShaderSource,
+        Texture,
+        Camera,
+        CullFace,
+        OrthographicFrustum,
+        Pass,
+        PerInstanceColorAppearance,
+        Primitive) {
+    "use strict";
+
+    // TODO : doc
+    // TODO : handle if depth texture extension is not supported
+
+    /**
+     * @private
+     */
+    function ShadowMap(scene) {
+        this.enabled = true;
+
+        this.debugShow = false;
+        this.debugFreezeFrame = false;
+        this._debugLightFrustum = undefined;
+        this._debugLight = undefined;
+        this._debugCameraFrustum = undefined;
+        this._debugCamera = undefined;
+        this._debugShadowViewCommand = undefined;
+
+        this._depthStencilTexture = undefined;
+        this._framebuffer = undefined;
+        this._size = 1024; // Width and height of the shadow map in pixels
+
+        this.renderState = RenderState.fromCache({
+            cull : {
+                enabled : true,
```

Comment:
You can just add a TODO, but I'm not sure that we can always backface cull here, e.g., think of a wall geometry casting a shadow.